src/server/snap.rs

```diff
@@ -110,96 +137,164 @@ impl Stream for SnapChunk {
 ///
 /// It will first send the normal raft snapshot message and then send the snapshot file.
 fn send_snap(
-    env: Arc<Environment>,
-    mgr: SnapManager,
-    security_mgr: Arc<SecurityManager>,
     addr: &str,
     msg: RaftMessage,
-) -> Result<()> {
+    env: Arc<Environment>,
+    security_mgr: Arc<SecurityManager>,
+    mgr: SnapManager,
+) -> Box<Future<Item = (), Error = Error> + Send> {
```

Comment:
can we use impl Traits to avoid Box here?

Reply:
This function won't be called frequently. And if we write it as `and_then` chain style, we need to clone many `Arc` or similar things. So I think readability is more important here.

Reply:
Why not use `Result<impl Future>`? So that you can use both `?` and `and_then`.